bson/src/main/org/bson/codecs/pojo/CollectionPropertyCodecProvider.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bson.codecs.pojo;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecConfigurationException;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+
+import static java.lang.String.format;
+
+final class CollectionPropertyCodecProvider implements PropertyCodecProvider {
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public <T> Codec<T> get(final TypeWithTypeParameters<T> type, final PropertyCodecRegistry registry) {
+        if (Collection.class.isAssignableFrom(type.getType())) {
+            if (type.getTypeParameters().size() == 0) {
+                throw new CodecConfigurationException("Invalid Collection type. Collections must have a defined type.");
```

Comment:
Previously, TypeData would have thrown an IllegalArgument Exception for a Collection without a defined type